Fb #9 ScheduleActivity

Luminilion opened the pull request:
The Travis test fails because we cannot build project because imports cannot be found in room.data.JoinedEvent. Otherwise, it should work with a hardcoded small demo of 2 concerts.
The functionnal way is set and ready but we need to find a way to get the list of Concerts from the current Event to the ScheduleActivity.

Commit:
Implemented basic schedule structure

Installed Timeline API and basic schedule structure to be customized
Timeline is composed of an  inflater, holder and associated layout files
+ dependencies : butterknife, cardView,recyclerView => app/build.graddle
modified

Commit:
Added Concert Class

Describes a concert occuring during the event. It is linked to the event
by a list contained by the event instance.
Concert has at this moment date, artist, genre and description fields.
Needed in the ScheduleActivity.

Commit:
Inclusion of the Concert list in Event

Added concertList as a List<Concert> in Event class.
Added a getter for the List.
Default list construction is LinkedList in main constructor.

Commit:
Added an addConcert method in the Event class.

Commit:
Redefined the equals method for it to properly compare 2 concerts

The redefinition of the method is necessary for the addConcert in the
Event class which checks if the list already contains a Concert that is
the same.

Commit:
Merge branches 'fb-#9' and 'master' of github.com:zyuiop/sdp-event-management into fb-#9

Commit:
Redesigned the TimeLine Cards to display all the Concert attributes

Added 2 TextViews, set default text for all fields.
Description Field is set by default to 40dp

Commit:
Adapted TimeLineModel for the Concert display

TimeLineModel transforms the Concert object into a Parcel used to
display the infos of the Concert (as far as I understood).
Changed the fields to include all those of Concert

Commit:
Adapted Schedule activity for it to display concerts for an event

No link to any event, this has to be set. 2 fake concerts manually
added to the dataList to be displayed.
Only missing thing is the automatic retrieval of the Concert List
associated to the current event.

Commit:
Set TimeLineAdapter class to properly integrate Concert objects

The Concert objects were transformed into a list of TimeLineModel which
contain the infos about the Concert.
Getters for each Concert field added in TimeLineModel.

These objects are put into TimeLineViewHolder objects.
TimeLineViewHolder attributes were linked to TextViews in layout files.

The TimeLineAdapter is the View equivalent (in the layout) to the List
containing the Concert objects. So it parses the TimeLineModel (which
are the transformed Concerts) into TimeLineViewHolder which will be
displayed.
Handling of this parsing was implemented in TimeLineAdapter.

Finally the ids of the views in the layout were renamed accordingly for
the binding.

Commit:
Added duration field to Concert and to layout

Commit:
Added isNow var and handled its initializiation

isNow describes the status of the concert, it is -1 if the concert
already happened, 0 if it is happening and 1 if it is going to happen.
Handled initialization and usage for timelineView

Commit:
Bound duration field to duration textview in layout

Commit:
Corrected typo in build.graddle

Commit:
Changes in TimeLineModel

Commit:
Further customized TimeLine object

Customized marker for each TimeLineView object depending on the status
of the Concert (playing now, already happened or will happen)
Integrated Duration display in the object
Added 3 marker states in the drawables

Commit:
Prepared Schedule activity for automatic fetching and displaying Concerts

For the moment only Concerts are hardcoded in this class. A list is ready
to hold all the Concerts from the event and to be displayed.

Commit:
Commited local version of TimeLineModel

Commit:
Messing arround with TimeLineAdapter/TimeLineModel and viewHolder

Commit:
Modified TimeLineAdapter to use String ressources instead of hardcoded values.

Luminilion committed:
Merge branch 'master' into fb-#9

Commit:
Modified TimeLineAdapter and concert to use String ressources instead of hardcoded values.

Commit:
Merge remote-tracking branch 'origin/fb-#9' into fb-#9

zyuiop committed:
Revert "Corrected typo in build.graddle"

This reverts commit f45be4c14c5477377354ece94f0593bf3a1857c9.

Commit:
Corrected bug in timeLineAdapter & TimeLineViewHolder. Changed LinearLayoutManager constructor in ScheduleActivity.

zyuiop reviewed (changes requested):
It crashes when running.

```
E/AndroidRuntime: FATAL EXCEPTION: main
                  Process: ch.epfl.sweng.eventmanager, PID: 14885
                  java.lang.IllegalStateException: Required view 'text_timeline_date' with ID 2131230906 for field 'mDate' was not found. If this view is optional add '@Nullable' (fields) or '@Optional' (methods) annotation.
                      at butterknife.internal.Utils.findRequiredView(Utils.java:92)
                      at butterknife.internal.Utils.findRequiredViewAsType(Utils.java:104)
                      at ch.epfl.sweng.eventmanager.ui.eventShowcase.scheduleUtils.TimeLineViewHolder_ViewBinding.<init>(TimeLineViewHolder_ViewBinding.java:22)
                      at java.lang.reflect.Constructor.newInstance0(Native Method)
                      at java.lang.reflect.Constructor.newInstance(Constructor.java:343)
                      at butterknife.ButterKnife.createBinding(ButterKnife.java:199)
                      at butterknife.ButterKnife.bind(ButterKnife.java:172)
                      at ch.epfl.sweng.eventmanager.ui.eventShowcase.scheduleUtils.TimeLineViewHolder.<init>(TimeLineViewHolder.java:35)
                      at ch.epfl.sweng.eventmanager.ui.eventShowcase.scheduleUtils.TimeLineAdapter.onCreateViewHolder(TimeLineAdapter.java:44)
                      at ch.epfl.sweng.eventmanager.ui.eventShowcase.scheduleUtils.TimeLineAdapter.onCreateViewHolder(TimeLineAdapter.java:23)
                      at android.support.v7.widget.RecyclerView$Adapter.createViewHolder(RecyclerView.java:6794)
                      at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:5975)
                      at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5858)
                      at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5854)
                      at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2230)
                      at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1557)
                      at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1517)
                      at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:612)
                      at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3924)
                      at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3641)
                      at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:4194)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.support.design.widget.CoordinatorLayout.layoutChild(CoordinatorLayout.java:1183)
                      at android.support.design.widget.CoordinatorLayout.onLayoutChild(CoordinatorLayout.java:870)
                      at android.support.design.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:889)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
                      at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.support.v7.widget.ActionBarOverlayLayout.onLayout(ActionBarOverlayLayout.java:444)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
                      at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
                      at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
                      at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
                      at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
                      at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
                      at com.android.internal.policy.DecorView.onLayout(DecorView.java:753)
E/AndroidRuntime:     at android.view.View.layout(View.java:20672)
                      at android.view.ViewGroup.layout(ViewGroup.java:6194)
                      at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2792)
                      at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2319)
                      at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1460)
                      at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7183)
                      at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
                      at android.view.Choreographer.doCallbacks(Choreographer.java:761)
                      at android.view.Choreographer.doFrame(Choreographer.java:696)
                      at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
                      at android.os.Handler.handleCallback(Handler.java:873)
                      at android.os.Handler.dispatchMessage(Handler.java:99)
                      at android.os.Looper.loop(Looper.java:193)
                      at android.app.ActivityThread.main(ActivityThread.java:6669)
                      at java.lang.reflect.Method.invoke(Native Method)
                      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

Commit:
Merge remote-tracking branch 'origin/fb-#9' into fb-#9

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/data/Concert.java:
```diff
@@ -0,0 +1,79 @@
+ * Class describing a single concert in an event. The class is for the moment only used by the
+ * ScheduleActivity.
+ */
+public final class Concert {
```
This class should probably not be named this way, as we don't only have concerts in our events. Should be something more generic, like `ScheduledEvent` or `ScheduledItem`

Luminilion replied:
Right this can be refactored during next Sprint.

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/data/Concert.java:
```diff
@@ -0,0 +1,79 @@
+            return compared.date.equals(date) && compared.artist.equals(artist) &&
+                    compared.genre.equals(genre) && compared.duration==duration;
+        } else return super.equals(obj);
+    }
```
Fix indentation

Luminilion replied:
There are no indentation problem, one return is in an if and the other is in the else.

zyuiop replied:
Yep mybad

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/data/Event.java:
```diff
@@ -29,13 +32,19 @@
         this.organizer = organizer;
         this.image = image;
+
+        this.concertList = new LinkedList<>();
```
Nope, it should come as an argument in the constructor.

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/data/Event.java:
```diff
@@ -57,4 +66,24 @@ public EventOrganizer getOrganizer() {
+     * @param concert Concert to be added to the List, can be null.
+     * @return true if the concert is in the list after returning.
+     */
+    public boolean addConcert(Concert concert) {
```
The idea of this class is to be imutable-ish, adding concerts breaks this goal. Remove this method.

Luminilion replied:
I implemented it as a reminder that the EventOrganizer can modify his event's schedule.

zyuiop replied:
We don't really know how we will do that

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/data/Event.java:
```diff
@@ -57,4 +66,24 @@ public EventOrganizer getOrganizer() {
     }
+
+    public List<Concert> getConcertList() {
+        return concertList;
```
Wrap with `Collections.unmodifiableList()`

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/ui/eventShowcase/ScheduleActivity.java:
```diff
@@ -1,15 +1,59 @@
+        //
+        // Get the concert List for the event into concertsList
+        //
+        concertsList = new LinkedList<>();
```
You added a concert list in `Event`, use that one!

Luminilion replied:
We did not decide during this Sprint how to pass the ConcertList to the ScheduleActivity. This is not the final code.

zyuiop replied:
Okay, but then why did you put Concert objects in the `Event` class?

zyuiop replied:
Removed.

zyuiop commented on app/src/main/java/ch/epfl/sweng/eventmanager/ui/eventShowcase/scheduleUtils/TimeLineModel.java:
```diff
@@ -0,0 +1,122 @@
+
+import java.util.Date;
+
+public class TimeLineModel implements Parcelable {
```
Why do we need a class like this? Can't we directly use `Concert` objects??

zyuiop replied:
Removed.

Commit:
Recommited the event_list_item.xml falsely deleted before

zyuiop reviewed (changes requested):
I think this PR misses some backend-logic to be ready to merge, and has some stuff that sounds duplicated.

For me, the lists of `Concert` should not be stored in the `Event` directly, but retrieved for somewhere else, as we don't want to make the `Event` too heavy (we pull them all when the app starts).

Commit:
Get concertList from constructor & return unmodifiableList

Luminilion committed:
Merge branch 'master' into fb-#9

zyuiop committed:
Fix compilation error

zyuiop committed:
Remove concerts from Event

zyuiop committed:
Create a repository for Concerts

zyuiop committed:
Refactor: move Schedule to its own module

zyuiop committed:
Remove TimeLineModel and move methods in Concert

zyuiop committed:
Create a ViewModel for the schedule and inject it

zyuiop committed:
Remove unused import

zyuiop committed:
Add a no-args constructor to Concert and use duration in minutes instead of hours

zyuiop committed:
Use firebase to store concerts

zyuiop committed:
Add a quick todo

zyuiop committed:
Fix lint errors

core-testing is a test library, it should be defined as an androidTestImplementation lib and not as an implementation list

zyuiop reviewed (commented):
Ok for me, but as I added some stuff at the end of the PR, I'd like someone else to review quickly.

Luminilion commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/ConcertRepository.java:
```diff
@@ -0,0 +1,53 @@
+ */
+@Singleton
+public class ConcertRepository {
+    private static String TAG = "ConcertRepository";
```
Should it not be in the strings.xml ? Otherwise, I think it should be final no ?

zyuiop replied:
Should be final yup, I let you change that

Luminilion commented on app/src/main/java/ch/epfl/sweng/eventmanager/repository/ConcertRepository.java:
```diff
@@ -0,0 +1,53 @@
+    public LiveData<List<Concert>> getConcerts(int eventId) {
+        final MutableLiveData<List<Concert>> data = new MutableLiveData<>();
+        DatabaseReference dbRef = FirebaseDatabase.getInstance()
+                .getReference("schedule_items")
```
Same as before, what strings go to strings.xml, as this one too is not going to change ?

zyuiop replied:
Strings that go to strings.xml are messages.
We might want to move these to constants, but not in strings.xml.

Luminilion reviewed (commented):
Seems all right to me but could we upgrade complexity to 6 ? 5 is very picky...

Luminilion committed:
Merge branch 'master' into fb-#9

zyuiop committed:
Make tag final
